### PR TITLE
Refactor grafana cli options

### PR DESCRIPTION
**Pull request by erickwilder**
### Checklist for a pull request to sitespeed.io

- [x] I'm making a big change or adding functionality so I've already opened an issue proposing the change to other contributors, so I got feedback on the idea before took the time to write precious code
- [x] Check that your change/fix has corresponding unit tests (if applicable)
- [x] Squash commits so it looks sane
- [x] ~~Update the documentation https://github.com/sitespeedio/sitespeed.io/tree/master/docs in another PR~~
- [x] Verify that the test works by running `npm test` and test linting by `npm run lint`

### Description
This is a proposition to start moving the responsibility of plugins defining their own CLI options. In an ideal situation _all_ plugins, not only built-in ones could benefit of such structure.
This is is related to #1065 - it doesn't address that issue entirely and I decided to put this to start a conversation about how process in next steps.

#### Suggestions

- Migrate all other plugins to match the proposed format (in separate PRs)
- Implement in the CLI some logic to load options from plugins added via command line/configuration (i.e `--plugins.add`)
- Update the plugin documentation to officially support `options()` method, _after_ we have a solid strategy to include 3rd party plugin CLI options

### CODE REVIEW

**Comment by soulgalore** [comment]
Hi @erickwilder ah cool.

I think the key issue (that stopped me to implement this before) is to find a good pattern your bullet number two:
>Implement in the CLI some logic to load options from plugins added via command line/configuration (i.e --plugins.add)

It would be good to iterate how that would work before we start move over to the new pattern.

**Comment by erickwilder** [comment]
I understand. I wanted to first work in the loading logic but as soon as I started digging in the cli code I saw that the codebase in there had a lot of distracting parts that could easily be moved over to the plug-in. The rationale here would be of cleaning up the code for cli _first_, then one could better focus on getting external plugins to define options.

Adding these options to the built-in plugins already has the benefit of a better separation of concerns. 

As for the loading strategy, I thought of separating the CLI processing in "phases" or "tiers" 

- Top-level options `--plugins.*` would be evaluated _before_ any other option
- Plugins listed in the configuration would probably have to be loaded in a similar way the `loadPlugins()` function work, but yielding just the plug-in module
- With a list of loaded plugins we could apply the concepts introduced in this PR to each one of them

I would ask you to consider moving the options to built-in plugins even without any dynamic loading of plugins. This is already the case for plugin authors: they have no way to add help messages in the standard Sitespeed help messages. Even if we have a new (for now undocumented) method, there is no regression but we pave the way to make that happen

**Comment by soulgalore** [comment]
Yep I see your point, sounds good. I also like the idea that each plugin handles the cli options.

File: lib/plugins/grafana/index.js [hunk 1]
```diff
@@ -1,11 +1,57 @@
...
+
 const defaultConfig = {
   port: 80
 };
+
 module.exports = {
+  name() {
+    return path.basename(__dirname);
+  },
+  options() {
```
Comment (soulgalore) [inline lib/plugins/grafana/index.js:15]: What about move those options out to a new file (cli.js or something) and set a new standard for that also? For some plugins the index.js is cluttered I'm thinking about minimising the code in there?

Comment (erickwilder) [reply lib/plugins/grafana/index.js]: I updated with the following:

* Renamed the proposed method to `cliOptions()` to be more explicit and to not shadow/override the property defined in the `.open()` method.
* Moved all definitions to `grafana/cli.js` that's later on `require`d by the method/property above

Not all plugins will have extensive options (like `browsertime`) and I think it's very convenient to keep the plugin interface clear with an explicit method/property like `cliOptions()`

**Comment by erickwilder** [comment]
@peter please, check my [latest commit](https://github.com/sitespeedio/sitespeed.io/pull/2984/commits/048cc71365d8595f666890161c9923fb7c104f5a). I took your suggestion into account _and_ decided to spend a bit more time to get some things prepared for upcoming PRs touching other plugins:

- Add some utility methods to `cli` so that it's easier to migrate built-in plugins to this format
- Add more tests to ensure nothing in the plugin interface (grafana only) related to these changes were broken
- Make `plugin.config` based on the yargs configuration (for all options defining an explicit `default:` property) instead of hard-coding them)

Let me know if this is good to go. Thanks!

**Comment by soulgalore** [comment]
Wow looks great @erickwilder !

**Pull request merged**